WIP: add hibernate orm (#24)
pingcap/dev-guide: en/for-hibernate-orm.md

---
title: App Development for Hibernate ORM
summary: Learn how to develop a simple Java application using TiDB and the Hibernate ORM.
---

# App Development for Hibernate ORM

This tutorial shows you how to build a simple Java application using TiDB and the Hibernate ORM, the best practices, and the known limitations.

## Step 1. Start a TiDB cluster

Start a pseudo TiDB cluster on your local storage:

```bash

> docker run pingcap/tidb:v5.1.0 -p 127.0.0.1:$LOCAL_PORT:4000

Reviewer: 恩，有道理，这个可以改一下...

```

The above command starts a temporary and single-node cluster with mock TiKV. The cluster listens on the port `$LOCAL_PORT`. After the cluster is stopped, any changes already made to the database are not persisted.

> **Note:**
>
> To deploy a "real" TiDB cluster for production, see the following guides:
>
> + [Deploy TiDB using TiUP for On-Premises](https://docs.pingcap.com/tidb/v5.1/production-deployment-using-tiup)
> + [Deploy TiDB on Kubernetes](https://docs.pingcap.com/tidb-in-kubernetes/stable)
>
> You can also [use TiDB Cloud](https://pingcap.com/products/tidbcloud/), a fully-managed Database-as-a-Service (DBaaS), which offers free trial.

## Step 2. Create a database

1. Connect to TiDB. Because TiDB is compatible with the MySQL protocol, you can connect to TiDB using any client. The following example uses the MySQL client:

    ```bash
    mysql -u root -h 127.0.0.1 -P $LOCAL_PORT
    ```

2. In the SQL shell, create the `bank` database that your application will use:

    ```sql
    CREATE DATABASE bank;
    ```

3. Create a SQL user for your application:

    ```sql
    CREATE USER <username> WITH PASSWORD <password>;
    ```

>     Take note of the username and password. You will use it in your application code later.

Author: ```suggestion
    Take note of the username and password. You will use them in your application code later.
```

   /**
    * Open session.
    *
    * @return the session
    */
   private static Session openSession() {
      if (sessionFactory == null) {
         final Configuration configuration = new Configuration();
         configuration.addAnnotatedClass(User.class);
         configuration.addAnnotatedClass(Order.class);

         sessionFactory = configuration.buildSessionFactory(new StandardServiceRegistryBuilder().build());
      }
      return sessionFactory.openSession();
   }
}
```

### Step 1. Get the application code

To get the `Example.java` code above, clone the `tidb-hibernate-example` repository to your machine:

```shell
git clone https://github.com/bb7133/tidb-hibernate-example
```

### Step 2. Update the connection parameters

Edit `src/main/resources/hibernate.properties` in a text editor:

1. Modify the `hibernate.connection.url` property with the port number from the `hibernate.properties` configuration file:

    ```
    hibernate.connection.url jdbc:mysql://127.0.0.1:4000/hibernate_example
    hibernate.connection.username root
    hibernate.connection.password
    ```

    In the above example, `4000` is the port number on which the TiDB cluster is listening.

> 2. Set the `hibernate.connection.username` property to the username you have created in [Create a database](#step-2-create-a-database).

Author: ```suggestion
2. Set the `hibernate.connection.username` property to the username you have created in [Create a database](#step-2-create-a-database).
```

> 2. Set the `hibernate.connection.username` property to the username you have created in [Create a database](#step-2-create-a-database).

Author: ```suggestion
2. Set the `hibernate.connection.username` property to the username you have created in [Create a database](#step-2-create-a-database).
```

2 rows in set (0.00 sec)

tidb> select * from Example$Order;                                                                               +---------+-------+--------+
+---------+-------+--------+
| orderId | price | userId |
+---------+-------+--------+
|       1 |   500 |      1 |
|       2 |   200 |      1 |
|       3 |   300 |      2 |
+---------+-------+--------+
```

If you try to delete the `Jack` row from `Example$User` table, using either the Hibernate code (`delete(session, jack)`) or the SQL statement (`delete from Example$User where userId=2;`), you might expect the an error as follows:

```sql
ERROR 1452 (23000): Cannot add or update a child row: a foreign key constraint fails (`hibernate_example`.`Example$Order`, CONSTRAINT `FKrd372ndoovvnmduu9iwffri3a` FOREIGN KEY (`orderId`) REFERENCES `Example$User` (`userId`))
```

However, the expected error is NOT reported by TiDB, because the foreign key constraint is ignored.

To avoid this situation, you need to maintain the mapping by yourself.

The support of foreign key is on the roadmap of TiDB. You can track the GitHub issue [#18209](https://github.com/pingcap/tidb/issues/18209) for more progress.

## What's next?

> Learn more about using the [Hibernate ORM](http://hibernate.org/orm/).

Author: ```suggestion
Learn more about how to use the [Hibernate ORM](http://hibernate.org/orm/).
```
